### DISCUSSION TITLE
Pathmind python reward terms

### DISCUSSION THREAD

**Discussion by slinlee** [PathmindAI/pathmind-webapp#3678] [Ideas]
Let's talk about how reward terms will be used in the webapp with Pathmind Python simulations. This is a proposal that I'm looking for feedback on.

cc @fionnachan @maxpumperla @ejunprung @alexamakarov @kepricon @chrisvnicholson @brettskymind @johnnyL7 @EvgeniyEA @SaharEs 

### Here is an example of a simple reward terms in a Pathmind model:
with two elements, `found_cheese` and `took_step`

```python 
 def get_reward(self, agent_id) -> typing.Dict[str, float]:
     return {
        "found_cheese": 1 if self.mouse == self.cheese else 0,
        "took_step": -1
     }
```

### **This is what the webapp's reward terms UI looks like for AnyLogic simulations:**

In AnyLogic models, users provide a list of variables they are interested in tracking. These are the simulation metrics or reward variables. These are often raw values from the simulation like 'goodsProduced' 'cost' 'throughput'

Users can pick from 'reward variables' to minimize or maximize. Behind the scenes, this generates a code snippet for each reward term like `reward = after.found_cheese - before.found_cheese;` This means we can also let users enter custom reward terms with java code and do basic validation that they're using variables that are available.

<img width="625" alt="Screen Shot 2021-10-29 at 2 52 28 PM" src="https://user-images.githubusercontent.com/1197406/139505593-9a3aa7a1-b674-42d4-abe3-3c1f10d4062e.png">

### **Here is what would change for Pathmind simulations:**

For Pathmind simulations, the code snippet is _already written by the users in their Python simulation._ Their names would be  displayed in the webapp so that they can (1) choose to include it, or not, and (2) add importance weights. **This is the limit for reward shaping without going back to the Python simulation.**

- there isn't a built in concept of 'before' and 'after' so we will need to provide examples of 'gather' type goals.
- since these reward terms are calculations based on metrics to be used in a reward function, **we might need to have a separate list of 'simulation metrics'** that are just the raw values so we can track and display the metrics charts and run MC for the users in the webapp.

<img width="625" alt="Screen Shot 2021-10-29 at 2 54 17 PM" src="https://user-images.githubusercontent.com/1197406/139505822-fa05fd79-6bcc-4bb2-bf04-fbcb21be7169.png">

1. Checkboxes to include the reward term
2. no need to indicate maximize/minimize
3. no ability to add new reward terms from the webapp.

Again, I appreciate any feedback or proposals if you were thinking of something different. Thanks!

**Comment by ejunprung** [comment]
This is assuming the rewards are auto-normalized already, right? Otherwise, the user needs to be mindful of the magnitude of each reward before assigning weights. 

But yeah, other than that, it makes sense to me.

**Reply by slinlee** [reply]
Yeah, ideally we have normalization sorted out nicely. but I think the weights are helpful even without it.

@ejunprung - btw, what do you think about the idea of writing the reward terms in the webapp in python more like how we do with AnyLogic?

**Reply by ghost** [reply]
All looks good to me.

The python user will be less dragged down by the clutter of AL PathmindHelper fields and the re-exporting of an AL .jar file. So while they experiment with training, defining a generous amount of custom reward terms in their python script won't feel too bad.

**Reply by ejunprung** [reply]
@slinlee Yeah, I'd prefer to write my rewards in Pathmind like how we do it with AnyLogic. It's a lot more efficient.

**Comment by slinlee** [comment]
@maxpumperla @brettskymind - I was in the process of adding tests for reward terms and weights and realized we don't have a good way to specify chosen reward terms. And the weights are an array that needs the reward terms to always be in the same order.

What do you think about switching it to this for setting the weights? The keys are the same from the get_reward function:
```python
weights = {
  "found_cheese": 10.0,
  "took_step": 1.0
}
```

If the user only wants to include `found_cheese` in the reward function, it would be:
```python
weights = {
  "found_cheese": 1.0,
}
```

The weights param is optional. If weights isn't included at all, all the terms are summed together with weight=1.0 as it does now.

**Reply by slinlee** [reply]
I'll have to check how this changes things for anylogic models.

**Reply by ghost** [reply]
I like this idea.

**Comment by SaharEs** [comment]
Looks good to me too! I am just confused about this part:
> - there isn't a built in concept of 'before' and 'after' so we will need to provide examples of 'gather' type goals.
> - since these reward terms are calculations based on metrics to be used in a reward function, we might need to have a separate list of 'simulation metrics' that are just the raw values so we can track and display the metrics charts and run MC for the users in the webapp.

So does this mean all the metrics have to be 'gather' based? What if a user wants to write a reward function like this;
``reward = Math.pow(after.distance, 2) ``
With no after-before. They wont be able to do this?

**Reply by slinlee** [reply]
- I don't think this means have to be gather based. It would be like anylogic sims where some are `found_cheese` or `landed`
- the reward terms are written in the python simulation code, so the user can do that if they want.

**Reply by SaharEs** [reply]
hmm assume a user defines this metric in their python simulation:
`metric_distance = -1 * math.pow(distance, 2)`

In the webapp when they choose this metric, behind the scene dont we generate a reward term like: `reward = after.metric_distance - before.metric_distance;` ? What if user wants the term to be `reward = after.metric_distance`?

**Reply by ghost** [reply]
@SaharEs I think @slinlee is saying that -- for pathmind simulation uploads -- without the maximize/minimize goal button, there will be no mapping to after-before. Instead the **reward term is taken as is** (except multiplied by a weight and possible normalization factor).

In their pathmind simulation, they may define a dictionary of many reward terms, **then choose which to use and emphasize through the webapp**:

```
def get_reward(self, agent_id) -> typing.Dict[str, float]:
     return {
         "distance": self.distance,
         "metric_distance": -1 * math.pow(self.distance, 2),
         "change_distance": self.distance - self.old_distance
         "change_metric_distance": ( -1 * math.pow(self.distance, 2) ) - ( -1 * math.pow(self.old_distance, 2) )
     }
```

And you **add multiplier weights to them in a separate dictionary with the same keys**, which can be overridden by the webapp.

**Reply by slinlee** [reply]
Yes, thanks for elaborating. In Pathmind simulations, we wouldn't generate an `after-before` term for them. We use what they give us.

**Reply by SaharEs** [reply]
@brettskymind Thanks for explaining. So the `get_reward` function is returning only the reward for a single step.

And in order to show the final value of metrics in the webapp, we will have to sum up the metrics throughout the episode? Is this a good assumption? This value will not have much meaning for metrics like `metric_distance`

**Reply by SaharEs** [reply]
I still think using `after-before` makes more sense. Plus, this way we can allow users to write custom reward terms in webapp which is much more efficient. So instead of defining all these metrics and manually holding account of `old_distance`, they would just give a `distance` value to pathmind. Wouldn't this add more value to our product?
```
         "distance": self.distance,
         "metric_distance": -1 * math.pow(self.distance, 2),
         "change_distance": self.distance - self.old_distance
         "change_metric_distance": ( -1 * math.pow(self.distance, 2) ) - ( -1 * math.pow(self.old_distance, 2) )
```

**Reply by ghost** [reply]
You bring up a good point. At each step, we could save a copy of the entire reward dictionary to characterize the "before step", and give them the option to shape rewards in webapp using this. Assuming python users want to use the webapp UI for this.

For final value of metrics in the webapp:
We track the episode contribution of each reward term, as well as the max, min, and mean of each reward term. (As of now, in pathmind simulations, there are just reward terms for metrics.) We need to decide what is necessary to show, is something missing from these metrics?

**Reply by SaharEs** [reply]
So webapp is a place of doing experimentation; defining new reward terms is a part of the experimentation. If someone prefers to define reward terms in their python script they still can. But we already have a nice UI that can do the job (possibly more efficiently), why retire that?

**Reply by slinlee** [reply]
Good point. Gym examples have the reward built in because they are used to compare algorithms against standard environments. Where our use case is experimenting with reward functions to get a good policy.

**Comment by maxpumperla** [comment]
@slinlee sorry for showing up late to the party. I'm a little confused by your argumentation, but hopefully we can clear that all up.

- First of all, why wouldn't we want to create `before` and `after` for Python simulations in the same way we do for AL? This is not difficult to pull off in nativerl and would be good for consistency (i.e. not having to modify our UI etc.). We're doing the exact same thing for AL models right now by storing `before = prev_reward` and `after = current_reward`. I don't see why we'd want to deviate from that pattern now (and I'm saying this as someone who's massively opposed to the whole before vs after shenanigans in the first place)? It's an established pattern, and I think we should keep it. It's proven to be valuable, e.g. for the reasons @SaharEs mentions above.
- After discussing with you (@slinlee), @brettskymind and @SaharEs, we decided to remove `get_metrics` from the Python simulation interface for two reasons. a) so that we don't confuse users and b) so that `get_reward` can act precisely like its AL equivalent, namely by providing reward variables / metrics / whatever you want to call it (btw, still not happy with the lack of consistent nomenclature around this topic).
- `get_reward` is a dictionary, meaning we provide "named values". That's my idea of a reward variable that can be used for shaping rewards in the Pathmind web UI.
- In the _absence of a reward term spec_, just sum up all provided terms. E.g. when starting training locally with `sim.train()` and forcing `start=True` this might be useful, as in this case users will use their Python environment for reward term crafting and kicking-off training (and not even touch the web UI).
- But I think we should encourage users to got to our UI and do the experimental setup there, for all the benefits like better parameter tracking etc.

**Reply by slinlee** [reply]
Thanks, @maxpumperla . I'm glad that I wrote this out then, even to throw away most of what I said in the original post!

1. Okay, let's build in `before` and `after`. I wasn't sure if you were still massively opposed to the idea :) I'll create an issue.
2. reward/metrics nomenclature - yes, it's a mess based on how the product go to where it is now. After this discussion I've started to think of it this way:
  - there is the `training` and there is the `environment`. The environment doesn't contain a reward function, the training experiment does. 
  
 
| training experiments (aka the webapp)  | environment (aka the simulation)|
| --- | --- |
| **reward function <- this is different from Gym env.** | logic of simulation |
| selecting obs | possible actions and making their effect |
| hyperparams x algorithms | **sharing important metrics, both to be used in reward function and to measure performance**|
| measuring and plotting metrics to show policy's performance.  | sharing observations available to agents |

So basically, we should switch to use the term `metrics` for what we gather in the simulation in both AnyLogic and Pathmind. 

- The Pathmind Helper labels them metrics now, but the classname is still `Reward`. This should be switched, along with any internal names that no longer make sense.
![](https://downloads.intercomcdn.com/i/o/264998970/da8b9d88785c71bf0c63a0d9/image.png)
- On the Python side, we will switch the name `get_reward` to `get_metrics`. 

Thoughts?

**Reply by SaharEs** [reply]
This all looks good to me!

**Reply by maxpumperla** [reply]
Great, thanks @slinlee. I'm not sure we can entirely disentangle envs and experiments, since it already breaks with observations (e.g. your field that says `sharing observations available to agents`).

On the Python side, I'd lean toward leaving things as they are for two reasons:

- rewards are natural for our intended target audience, metrics might be confusing
- when users go down the "sum all things up" route, you'd rather have an intuition of summing up all items in the reward function.

I'm more concerned with a consistent interface in the web app than I am about streamlining Python vs AL. In the end, I'd assume the intersection between those two groups will be close to 0 in practice.

**Reply by maxpumperla** [reply]
oh, and on a practical level it won't be easy to simultaneously release a new pathmind API and nativerl version that switches to `get_metrics`. we can do that, though. not totally opposed to it.

**Reply by slinlee** [reply]
ok. So on the python side what do you think of 
1) emphasizing `get_metrics` for webapp users
2) the webapp uses get_metrics to create a new reward function, using the existing functionality to override reward functions.

**Reply by maxpumperla** [reply]
for that we'd have to reintroduce `get_metrics` again, but should we? I feel like we can get by with just one function that serves different purposes, as long as we have good documentation. otherwise we have to explain when to use which yet again.

**Reply by slinlee** [reply]
How would we: 
1) plot metrics and run monte carlo in the webapp?
2) also let users write reward terms in the webapp?

if those are covered with the one function, then I guess it works.

**Reply by SaharEs** [reply]
I don't think we should assume that our users will be RL experts who rewards are natural to them. Even if they are, changing the name to get_metrics would be a good signal for them to think about it differently.

**Reply by maxpumperla** [reply]
I thought all that was covered by one function in AL, too. Is that not the case?

**Reply by slinlee** [reply]
In AL, we get the metrics. When we train the reward function is a separate 'reward snippet' passed into the train.sh script. It's separate from the metrics being passed in.

**Reply by maxpumperla** [reply]
exactly, so we don't have two separate things in AL either. that's my point. whether or not we rename rewards to metrics or vice versa does not make that much of a difference.

**Reply by slinlee** [reply]
Ok. But that's not implemented now, right? It doesn't accept a separate reward function that doesn't override get_reward and this just sums up the terms https://github.com/SkymindIO/nativerl/blob/dev/nativerl/python/pathmind_training/environments.py#L440

Let me check on this today. If it's there already, that's great! If it was removed, we can add it again.

If we can do the two things I listed before, then I'm happy for now. The terminology confusion will just stay in the back log.

> 1. plot metrics and run monte carlo in the webapp?
> 2. also let users write reward terms in the webapp?
